pkg/sql/sqlbase/testutils.go

```diff
@@ -847,40 +853,106 @@ func RandCreateTables(
+func RandCreateTableWithInterleave(
+	rng *rand.Rand, prefix string, tableIdx int, interleaveInto *tree.CreateTable,
+) *tree.CreateTable {
 	// columnDefs contains the list of Columns we'll add to our table.
-	columnDefs := make([]*tree.ColumnTableDef, randutil.RandIntInRange(rng, 1, 20))
+	nColumns := randutil.RandIntInRange(rng, 1, 20)
+	columnDefs := make([]*tree.ColumnTableDef, 0, nColumns)
 	// defs contains the list of Columns and other attributes (indexes, column
 	// families, etc) we'll add to our table.
-	defs := make(tree.TableDefs, len(columnDefs))
-
-	for i := range columnDefs {
-		columnDef := randColumnTableDef(rng, i)
-		columnDefs[i] = columnDef
-		defs[i] = columnDef
+	defs := make(tree.TableDefs, 0, len(columnDefs))
+
+	// Find columnDefs from previous create table.
+	interleaveIntoColumnDefs := make(map[tree.Name]*tree.ColumnTableDef)
+	var interleaveIntoPK *tree.UniqueConstraintTableDef
+	if interleaveInto != nil {
+		for i := range interleaveInto.Defs {
+			if cd, ok := interleaveInto.Defs[i].(*tree.ColumnTableDef); ok {
+				interleaveIntoColumnDefs[cd.Name] = cd
+			}
+			if uc, ok := interleaveInto.Defs[i].(*tree.UniqueConstraintTableDef); ok && uc.PrimaryKey {
+				interleaveIntoPK = uc
+			}
+		}
 	}
+	var interleaveDef *tree.InterleaveDef
+	if interleaveIntoPK != nil && len(interleaveIntoPK.Columns) > 0 {
+		// Make the interleave prefix, which has to be exactly the columns in the
+		// parent's primary index.
+		prefixLength := len(interleaveIntoPK.Columns)
+		fields := make(tree.NameList, prefixLength)
+		for i := range interleaveIntoPK.Columns[:prefixLength] {
+			fmt.Println(interleaveIntoPK.Columns[i].Column, interleaveIntoColumnDefs)
+			def := interleaveIntoColumnDefs[interleaveIntoPK.Columns[i].Column]
+			columnDefs = append(columnDefs, def)
+			defs = append(defs, def)
+			fields[i] = def.Name
+		}
 
-	// Shuffle our column definitions.
-	rng.Shuffle(len(columnDefs), func(i, j int) {
-		columnDefs[i], columnDefs[j] = columnDefs[j], columnDefs[i]
-	})
+		extraCols := make([]*tree.ColumnTableDef, nColumns)
+		// Add more columns to the table.
+		for i := range extraCols {
+			extraCol := randColumnTableDef(rng, tableIdx, i+prefixLength)
+			extraCols[i] = extraCol
+			columnDefs = append(columnDefs, extraCol)
+			defs = append(defs, extraCol)
+		}
 
-	// Make a random primary key with high likelihood.
-	if rng.Intn(8) != 0 {
-		indexDef := randIndexTableDefFromCols(rng, columnDefs)
-		if len(indexDef.Columns) > 0 {
-			defs = append(defs, &tree.UniqueConstraintTableDef{
-				PrimaryKey:    true,
-				IndexTableDef: indexDef,
+		rng.Shuffle(nColumns, func(i, j int) {
+			extraCols[i], extraCols[j] = extraCols[j], extraCols[i]
+		})
+
+		// Create the primary key to interleave, maybe add some new columns to the
+		// one we're interleaving.
+		pk := &tree.UniqueConstraintTableDef{
+			PrimaryKey: true,
+			IndexTableDef: tree.IndexTableDef{
+				Columns: interleaveIntoPK.Columns[:prefixLength],
```

Comment:
add another `:prefixLength` here to ensure that any mutation to this doesn't mutate the underlying slice